File: backend/app/schemas/bill.py
"""Bill reminder schemas for API validation."""

from datetime import date, datetime, time
from typing import Any
from pydantic import BaseModel, Field, field_validator


class BillCreate(BaseModel):
    """Schema for creating a bill."""

    name: str = Field(..., min_length=1, max_length=255)
    description: str | None = None
    amount: int = Field(..., gt=0)
    category: str = Field(..., min_length=1, max_length=100)
    color: str | None = Field(None, pattern="^#[0-9A-Fa-f]{6}$")
    due_day: int = Field(..., ge=1, le=31)
    due_time: time | None = None
    is_recurring: bool = True
    recurrence_type: str = Field(
        default="monthly", pattern="^(weekly|biweekly|monthly|quarterly|yearly|custom)$"
    )
    recurrence_config: dict[str, Any] | None = None
    next_due_date: date
    last_paid_date: date | None = None
    reminder_days_before: int = Field(default=3, ge=1, le=7)
    reminder_enabled: bool = True
    recurring_transaction_id: int | None = None
    sync_with_recurring: bool = True

    @field_validator("next_due_date")
    @classmethod
    def validate_next_due_date(cls, v: date) -> date:
        if v < date.today():
            raise ValueError("next_due_date cannot be in the past")
        return v


class BillUpdate(BaseModel):
    """Schema for updating a bill."""

    name: str | None = None
    description: str | None = None
    amount: int | None = Field(None, gt=0)
    category: str | None = None
    color: str | None = Field(None, pattern="^#[0-9A-Fa-f]{6}$")
    due_day: int | None = Field(None, ge=1, le=31)
    due_time: time | None = None
    is_recurring: bool | None = None
    recurrence_type: str | None = Field(
        None, pattern="^(weekly|biweekly|monthly|quarterly|yearly|custom)$"
    )
    recurrence_config: dict[str, Any] | None = None
    next_due_date: date | None = None
    last_paid_date: date | None = None
    reminder_days_before: int | None = Field(None, ge=1, le=7)
    reminder_enabled: bool | None = None
    sync_with_recurring: bool | None = None


class BillHistoryResponse(BaseModel):
    """Bill payment history response."""

    id: int
    bill_id: int
    paid_date: date
    amount_paid: int
    notes: str | None = None
    created_at: datetime

    class Config:
        from_attributes = True


class BillResponse(BaseModel):
    """Bill response."""

    id: int
    user_id: int
    name: str
    description: str | None = None
    amount: int
    category: str
    color: str | None = None
    due_day: int
    due_time: time | None = None
    is_recurring: bool
    recurrence_type: str
    recurrence_config: dict[str, Any] | None = None
    next_due_date: date
    last_paid_date: date | None = None
    reminder_days_before: int
    reminder_enabled: bool
    last_reminder_sent: datetime | None = None
    is_paid: bool
    paid_amount: int | None = None
    recurring_transaction_id: int | None = None
    sync_with_recurring: bool
    last_synced_at: datetime | None = None
    created_at: datetime
    updated_at: datetime
    is_active: bool

    class Config:
        from_attributes = True


class BillListResponse(BaseModel):
    """Response for list of bills."""

    success: bool = True
    bills: list[BillResponse]
    total_count: int


class BillCalendarDay(BaseModel):
    """Single day in calendar view."""

    day: int
    bills: list[BillResponse]


class BillCalendarResponse(BaseModel):
    """Response for bill calendar view."""

    success: bool = True
    year: int
    month: int
    days: list[BillCalendarDay]
    total_bills_due: int
    total_amount_due: int


class UpcomingBillResponse(BaseModel):
    """Upcoming bill response."""

    id: int
    name: str
    amount: int
    due_date: date
    days_until_due: int
    category: str
    reminder_enabled: bool
    reminder_days_before: int


class UpcomingBillsResponse(BaseModel):
    """Response for upcoming bills."""

    success: bool = True
    upcoming_bills: list[UpcomingBillResponse]
    total_count: int
    total_amount: int


class MarkBillPaidRequest(BaseModel):
    """Request to mark bill as paid."""

    paid_date: date | None = None
    amount_paid: int | None = None
    notes: str | None = None


class MarkBillPaidResponse(BaseModel):
    """Response for marking bill as paid."""

    success: bool = True
    id: int
    is_paid: bool
    paid_date: date
    next_due_date: date
    history_id: int


class BillDetailResponse(BaseModel):
    """Bill detail with history response."""

    success: bool = True
    bill: BillResponse
    history: list[BillHistoryResponse]


class BillDeleteResponse(BaseModel):
    """Response for deleting a bill."""

    success: bool = True
    message: str = "Bill deleted successfully"


# === Reminder Schedule Schemas ===


class ReminderScheduleCreate(BaseModel):
    """Schema for creating a reminder schedule."""

    reminder_type: str = Field(
        default="days_before", pattern="^(days_before|specific_date|recurring)$"
    )
    days_before: int | None = Field(None, ge=1, le=14)
    reminder_time: time | None = Field(None, description="Time of day for reminder (HH:MM)")
    recurrence_config: dict[str, Any] | None = None


class ReminderScheduleResponse(BaseModel):
    """Reminder schedule response."""

    id: int
    bill_id: int
    reminder_type: str
    days_before: int | None
    reminder_time: datetime
    is_sent: bool
    sent_at: datetime | None
    created_at: datetime

    class Config:
        from_attributes = True


class ReminderScheduleListResponse(BaseModel):
    """Response for list of reminder schedules."""

    success: bool = True
    schedules: list[ReminderScheduleResponse]


class ReminderScheduleDeleteResponse(BaseModel):
    """Response for deleting a reminder schedule."""

    success: bool = True
    message: str = "Reminder schedule deleted successfully"


# === Partial Payment Schemas ===


class PartialPaymentStatusResponse(BaseModel):
    """Partial payment status response."""

    success: bool = True
    bill_id: int
    total_amount: int
    paid_amount: int
    remaining_amount: int
    is_fully_paid: bool
    has_partial_payment: bool

File: frontend/src/types/bill.ts
export interface Bill {
  id: number;
  user_id: number;
  name: string;
  description: string | null;
  amount: number;
  category: string;
  color: string | null;
  due_day: number;
  due_time: string | null;
  is_recurring: boolean;
  recurrence_type: 'weekly' | 'biweekly' | 'monthly' | 'quarterly' | 'yearly' | 'custom' | 'none';
  recurrence_config: Record<string, unknown> | null;
  next_due_date: string;
  last_paid_date: string | null;
  reminder_days_before: number;
  reminder_enabled: boolean;
  last_reminder_sent: string | null;
  is_paid: boolean;
  paid_amount: number | null;
  recurring_transaction_id: number | null;
  sync_with_recurring: boolean;
  last_synced_at: string | null;
  created_at: string;
  updated_at: string;
  is_active: boolean;
  notes?: string;
}

export interface BillHistory {
  id: number;
  bill_id: number;
  paid_date: string;
  amount_paid: number;
  notes: string | null;
  created_at: string;
}

export interface BillCreate {
  name: string;
  description?: string;
  amount: number;
  category: string;
  color?: string;
  due_day: number;
  due_time?: string;
  is_recurring?: boolean;
  recurrence_type?: 'weekly' | 'biweekly' | 'monthly' | 'quarterly' | 'yearly' | 'custom' | 'none';
  recurrence_config?: Record<string, unknown>;
  next_due_date: string;
  last_paid_date?: string;
  reminder_days_before?: number;
  reminder_enabled?: boolean;
  recurring_transaction_id?: number;
  sync_with_recurring?: boolean;
  is_paid?: boolean;
  notes?: string;
}

export interface BillUpdate {
  name?: string;
  description?: string;
  amount?: number;
  category?: string;
  color?: string;
  due_day?: number;
  due_time?: string;
  is_recurring?: boolean;
  recurrence_type?: 'weekly' | 'biweekly' | 'monthly' | 'quarterly' | 'yearly' | 'custom';
  recurrence_config?: Record<string, unknown>;
  next_due_date?: string;
  last_paid_date?: string;
  reminder_days_before?: number;
  reminder_time?: string;
  reminder_enabled?: boolean;
  sync_with_recurring?: boolean;
}

export interface BillCalendarDay {
  day: number;
  bills: Bill[];
}

export interface BillCalendarResponse {
  success: boolean;
  year: number;
  month: number;
  days: BillCalendarDay[];
  total_bills_due: number;
  total_amount_due: number;
}

export interface UpcomingBill {
  id: number;
  name: string;
  amount: number;
  due_date: string;
  days_until_due: number;
  category: string;
  reminder_enabled: boolean;
  reminder_days_before: number;
  is_paid?: boolean;
}

export interface UpcomingBillsResponse {
  success: boolean;
  upcoming_bills: UpcomingBill[];
  total_count: number;
  total_amount: number;
}

export interface BillListResponse {
  success: boolean;
  bills: Bill[];
  total_count: number;
}

export interface MarkBillPaidRequest {
  paid_date?: string;
  amount_paid?: number;
  notes?: string;
}

export interface MarkBillPaidResponse {
  success: boolean;
  id: number;
  is_paid: boolean;
  paid_date: string;
  next_due_date: string;
  history_id: number | null;
}

// === Reminder Schedule Types ===

export interface ReminderSchedule {
  id: number;
  bill_id: number;
  reminder_type: 'days_before' | 'specific_date' | 'recurring';
  days_before: number | null;
  reminder_time: string;
  is_sent: boolean;
  sent_at: string | null;
  created_at: string;
}

export interface ReminderScheduleCreate {
  reminder_type?: 'days_before' | 'specific_date' | 'recurring';
  days_before?: number;
  reminder_time?: string; // HH:MM format
  recurrence_config?: Record<string, unknown>;
}

export interface ReminderScheduleListResponse {
  success: boolean;
  schedules: ReminderSchedule[];
}

export interface ReminderScheduleDeleteResponse {
  success: boolean;
  message: string;
}

// === Partial Payment Types ===

export interface PartialPaymentStatusResponse {
  success: boolean;
  bill_id: number;
  total_amount: number;
  paid_amount: number;
  remaining_amount: number;
  is_fully_paid: boolean;
  has_partial_payment: boolean;
}
